File: java/src/main/java/com/mlt/vector/BitVector.java
package com.mlt.vector;

import java.nio.ByteBuffer;

public class BitVector {

  private final ByteBuffer values;
  private final int size;

  // TODO: check how BitSet is ordered
  /**
   * @param values The byte buffer containing the bit values in least-significant bit (LSB)
   *     numbering
   */
  public BitVector(ByteBuffer values, int size) {
    this.values = values;
    this.size = size;
  }

  public boolean get(int index) {
    int byteIndex = index / 8;
    int bitIndex = index % 8;
    byte b = values.get(byteIndex);
    return ((b >> bitIndex) & 1) == 1;
  }

  public int getInt(int index) {
    int byteIndex = index / 8;
    int bitIndex = index % 8;
    byte b = values.get(byteIndex);
    return (b >> bitIndex) & 1;
  }

  public int size() {
    return size;
  }

  public ByteBuffer getBuffer() {
    return this.values;
  }
}

File: ts/src/vector/flat/bitVector.ts
export default class BitVector {
    private readonly values: Uint8Array;
    private readonly _size: number;

    /**
     * @param values The byte buffer containing the bit values in least-significant bit (LSB)
     *     numbering
     */
    constructor(values: Uint8Array, size: number) {
        this.values = values;
        this._size = size;
    }

    get(index: number): boolean {
        const byteIndex = Math.floor(index / 8);
        const bitIndex = index % 8;
        const b = this.values[byteIndex];
        return ((b >> bitIndex) & 1) === 1;
    }

    set(index: number, value: boolean): void{
        //TODO: refactor -> improve quick and dirty solution
        const byteIndex = Math.floor(index / 8);
        const bitIndex = index % 8;
        this.values[byteIndex] = this.values[byteIndex] | ((value ? 1 : 0) << bitIndex);
    }

    getInt(index: number): number {
        const byteIndex = Math.floor(index / 8);
        const bitIndex = index % 8;
        const b = this.values[byteIndex];
        return (b >> bitIndex) & 1;
    }

    size(): number {
        return this._size;
    }

    getBuffer(): Uint8Array {
        return this.values;
    }
}
